### ISSUE TITLE
Require options to be an object

### ISSUE THREAD

**Issue by sffc** [tc39/ecma402#480]
Currently, we coerce the options argument to an object by calling `ToObject`.  @justingrant suggested in https://github.com/tc39/proposal-temporal/pull/693 that we should be stricter and throw instead.

I don't think we should change older Intl objects, due to backwards compatibility, but we can adopt this pattern in new objects.

**Comment by anba** [comment]
If there's agreement to perform this change, we should definitely also try to apply it to the existing Intl constructors. And only if we detect any web-compat issues, back it out for the affected constructors. Similar to how we managed to restrict the "IntlLegacyConstructedSymbol" hack to just Intl.NumberFormat and Intl.DateTimeFormat, but not Intl.Collator.

**Comment by ptomato** [comment]
There's a way to make this change non-normatively as well. We could change the definition of GetOption but make sure to use ToObject first on all objects passed to existing uses of GetOption. That way, the definition of GetOption could move into 262 when Temporal does, but the behaviour in Intl could remain unchanged.

I'll prepare a PR with both the editorial and normative changes, split into two commits, for discussion.

**Comment by ptomato** [comment]
OK, this is in #493, one non-normative commit followed by a normative commit.

**Comment by sffc** [comment]
Another option, which I mentioned in the meeting:

Keep `GetOption` defined as it is now, and move it to 262.  In 262, make new abstract operations as follows:

```html
<emu-clause id="sec-getoption" aoid="NormalizeOptionsObject">
  <h1>NormalizeOptionsOption ( _options_, _legacyMode_ )</h1>
  <p>
    The abstract operation NormalizeOptionsObject massages _options_ into an Object to be subsequently passed to GetOption.
  </p>
  <emu-alg>
    1. If _options_ is *undefined*, then
      1. Return ObjectCreate(*null*).
    1. If Type(_options_) is Object, then
      1. Return _options_.
    1. If _legacyMode_, then
      1. Return ToObject(_options_).
    1. Throw a *TypeError* exception.
  </emu-alg>
</emu-clause>
```

Change the 402 call sites as follows:

```html
<!-- Current 4 lines, present in several places, preceding calls to GetOption -->
1. If _options_ is *undefined*, then	
  1. Let _options_ be ObjectCreate(*null*).	
1. Else,	
  1. Let _options_ be ? ToObject(_options_).

<!-- Replace with: -->
1. Let _options_ be NormalizeOptionsObject(_options_, true).
```

All new callers of `NormalizeOptionsObject` should have legacy mode disabled. We would enable it for the existing 402 call sites only so that the change is purely editorial.

@jswalden @FrankYFTang @gibson042 @leobalter

**Comment by ljharb** [comment]
Rather than adding a "legacy mode" into 262, why not have an operation in 402 that calls NormalizeOptionsOption, and if it's abrupt, wraps the original value in ToObject?

**Comment by sffc** [comment]
If we don't want the legacy mode in 262, we could do:

```html
<!-- 262 -->
<emu-clause id="sec-normalizeoptionsobject" aoid="NormalizeOptionsObject">
  <h1>NormalizeOptionsObject ( _options_ )</h1>
  <p>
    The abstract operation NormalizeOptionsObject massages _options_ into an Object to be subsequently passed to GetOption. It throws a TypeError if _options_ is not undefined and not an Object.
  </p>
  <emu-alg>
    1. If _options_ is *undefined*, then
      1. Return ObjectCreate(*null*).
    1. If Type(_options_) is Object, then
      1. Return _options_.
    1. Throw a *TypeError* exception.
  </emu-alg>
</emu-clause>

<!-- 402 -->
<emu-clause id="sec-normalizeoptionsobjectlegacy" aoid="NormalizeOptionsObjectLegacy">
  <h1>NormalizeOptionsObjectLegacy ( _options_ )</h1>
  <p>
    The abstract operation NormalizeOptionsObjectLegacy massages _options_ into an Object to be subsequently passed to GetOption. It does not throw an exception, instead making a best-effort attempt at coercing _options_ to an Object. Newer spec code should use ECMA-262's NormalizeOptionsObject instead.
  </p>
  <emu-alg>
    1. If _options_ is *undefined*, then
      1. Return ObjectCreate(*null*).
    1. If Type(_options_) is Object, then
      1. Return _options_.
    1. Return ToObject(_options_).
  </emu-alg>
</emu-clause>
```

**Comment by ptomato** [comment]
NormalizeOptionsObject and NormalizeOptionsObjectLegacy seem reasonable to me. To get things moving, I would propose adding them both to the 402 spec, then we can later move NormalizeOptionsObject and GetOption into 262?

**Comment by FrankYFTang** [comment]
I support @ptomato's suggestion in  https://github.com/tc39/ecma402/issues/480#issuecomment-673703900 and @sffc  proposal in https://github.com/tc39/ecma402/issues/480#issuecomment-679420266

**Comment by ptomato** [comment]
I've updated the PR so that it doesn't make any normative changes. Please take a look when you have time!

**Comment by ptomato** [comment]
As per today's TC39 meeting, these are the abstract operations that would be moved into ECMA-262 as part of Temporal:
- NormalizeOptionsObject: https://github.com/tc39/ecma402/pull/493/files#diff-8f6d7f17f474d8216e2f15477e3a926dR306-R319
- GetOption: https://github.com/tc39/ecma402/pull/493/files#diff-8f6d7f17f474d8216e2f15477e3a926dR336-R355

@erights Are these definitions sufficient / compatible for the other proposals that you mentioned in the meeting (compartments, realms, ...)?

@littledan @ljharb If we want to discuss `null` or not `null` this would be the place to do it!

**Comment by ljharb** [comment]
I think `null` should throw, as should any non-undefined non-object.

**Comment by erights** [comment]
We will investigate and discuss. Thanks!

Attn @kriskowal @fudco @bmeck @phoddie @patrick-soquet @kumavis

**Comment by littledan** [comment]
I don't have a strong opinion on null vs undefined here. I'm happy to leave things as they are.

**Comment by kriskowal** [comment]
> We will investigate and discuss. Thanks!
> 
> Attn @kriskowal @FUDCo @bmeck @phoddie @patrick-soquet @kumavis @caridy

I do not have strong opinions. We should be able to make good use of any such utilities for upcoming proposals with options bags, like `Compartment`. I don’t believe we have any use for distinguishing `null` from `undefined`.

**Comment by ptomato** [comment]
Any subsequent comments on this, either on the PR itself, or on potential compatibility issues with other proposals?

**Comment by ptomato** [comment]
I'm not available for the 402 monthly meeting tomorrow to discuss this, but if there's anything blocking merging the editorial PR, please let me know!

**Comment by sffc** [comment]
2021-02-11 discussion: https://github.com/tc39/ecma402/blob/master/meetings/notes-2021-02-11.md#require-options-to-be-an-object-480

We agreed to adopt the new stricter GetOptions behavior in newer APIs, but not to change older APIs to avoid breaking the web with what would most likely be benign errors.  We should document this decision by ensuring that new constructors use the new behavior.

**Comment by ljharb** [comment]
Is it not worth some telemetry to figure out if it would break the web?

**Comment by sffc** [comment]
> Is it not worth some telemetry to figure out if it would break the web?

I'd be happy if someone can volunteer to do this, but from a prioritization standpoint I don't want this to delay more urgent issues that are on our backlog.

**Comment by FrankYFTang** [comment]
> 2021-02-11 discussion: https://github.com/tc39/ecma402/blob/master/meetings/notes-2021-02-11.md#require-options-to-be-an-object-480
> 
> We agreed to adopt the new stricter GetOptions behavior in newer APIs, but not to change older APIs to avoid breaking the web with what would most likely be benign errors. We should document this decision by ensuring that new constructors use the new behavior.

Could you clearly define which one are part of the "newer APIs" and which one are part of the "older APIs"? The word "newer" and "older" is very ambiguous here.

**Comment by gibson042** [comment]
Any API introduced to the specification from this point forward, minus possible exceptions for anything that is already mature enough for that change itself to present a risk.

**Comment by sffc** [comment]
I filed https://github.com/tc39/ecma402/issues/537 to track updating new ES 2021 proposals.

**Comment by FrankYFTang** [comment]
FYI v8 tracking https://bugs.chromium.org/p/v8/issues/detail?id=11466
test262 tracking https://github.com/tc39/test262/issues/2950

**Comment by Constellation** [comment]
JSC change FYI https://github.com/WebKit/WebKit/commit/51b6659009f49c33837e1ed10cddbd454315fb44

**Comment by sffc** [comment]
Given that #538 and https://github.com/tc39/proposal-intl-segmenter/pull/135 are merged, is there anything remaining to do on this issue?

**Comment by gibson042** [comment]
Yes: https://github.com/tc39/ecma402/issues/480#issuecomment-777791334
> We should document this decision by ensuring that new constructors use the new behavior.

I'm working on similar changes now, and do expect this to close soon.
